### PR TITLE
APEXCORE-294 controlled shutdown of an application.

### PR DESCRIPTION
**Pull request by tushargosavi**

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/engine/Node.java [hunk 1]
```diff
@@ -317,9 +317,17 @@ public void removeSinks(Map<String, Sink<Object>> sinks)
...
   protected volatile boolean shutdown;
 
-  public void shutdown()
+  /**
+   * If shutdown the node by setting alive to false. If processEndWindow is true, then after shutting
+   * down endWindow is called and END_STREAM tuple is sent to downstream operators.
+   *
+   * @param processEndWindow
+   */
+  public void shutdown(boolean processEndWindow)
```
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/engine/Node.java]: How does an user specify that processEndWindow=true ?

File: engine/src/test/java/com/datatorrent/stram/engine/GenericNodeTest.java [hunk 2]
```diff
@@ -602,7 +602,7 @@ public void run()
       interval += sleeptime;
     } while ((gn.controlTupleCount == controlTupleCount) && (interval < maxSleep));
 
-    gn.shutdown();
+    gn.shutdown(false);
     t.join();
```
Comment (sandeshh) [inline engine/src/test/java/com/datatorrent/stram/engine/GenericNodeTest.java]: New test case is required to test processEndWindow = true?

File: engine/src/main/java/com/datatorrent/stram/engine/Node.java (hunk 1 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/engine/Node.java]: when user triggers dag shutdown operator through rest API, shutdown is called with processEndWindow = true. In all other cases processEndWindow is set false to preserve old behaviour.

File: engine/src/test/java/com/datatorrent/stram/engine/GenericNodeTest.java (hunk 2 above)
Comment (tushargosavi) [reply engine/src/test/java/com/datatorrent/stram/engine/GenericNodeTest.java]: I will add a new test case for this.

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java [hunk 3]
```diff
@@ -2260,6 +2260,29 @@ public void shutdownAllContainers(String message)
     }
   }
 
+  /**
+   * Send STOP_OPERATOR command to all input operator to initiate shutdown of a DAG.
+   *
+   * @param message
+   */
+  public void stopInputOperators(String message)
```
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: HeartBeat already can send Shutdown notification. Why introduce another method?

Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: engine can shutdown container in various cases, the shutdown field in heartbeat is used to terminate the container in error cases such as blocked operator

This is for user initiated shutdown, also the shutdown field is for whole container, we want to start termination from input operators.

Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: We can change the *shutdown* variable in the heartbeat to Enum to represent different ways of shutting the containers and operators. Idea is that all the shutdown logic is contained in one place.

Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: @sandeshh I have updated the code using shutdown field.

**Comment by tushargosavi** [comment]
@PramodSSImmaneni @sandeshh  please review

File: engine/src/main/java/com/datatorrent/stram/api/StreamingContainerUmbilicalProtocol.java [hunk 4]
```diff
@@ -341,6 +341,13 @@ public String toString()
     }
   }
 
+  enum ShutdownType
```
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/api/StreamingContainerUmbilicalProtocol.java:344]: Just Shutdown is fine.

File: engine/src/main/java/com/datatorrent/stram/api/StreamingContainerUmbilicalProtocol.java [hunk 5]
```diff
@@ -341,6 +341,13 @@ public String toString()
...
+  {
+    NO_SHUTDOWN,
```
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/api/StreamingContainerUmbilicalProtocol.java]: SHUTDOWN is redundant.

File: engine/src/main/java/com/datatorrent/stram/engine/Node.java (hunk 1 above)
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/engine/Node.java]: When a user triggers a shutdown, what should be the application state in Yarn? Killed or Finished?

Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/engine/Node.java]: The state of the application will be Finished.

Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/engine/Node.java]: You will have to make changes in StreamingAppMasterService, *forcedShutdown* variable will mark the application as *failed* instead of *Finished*

Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/engine/Node.java]: My thinking is users should decide whether the application is failed or succeeded.

https://issues.apache.org/jira/browse/APEXCORE-551

Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/engine/Node.java]: Current behavior is when application is shutdown its state is Finished. as you suggested this could be taken up through APEXCORE-551

File: engine/src/main/java/com/datatorrent/stram/api/StreamingContainerUmbilicalProtocol.java [hunk 6]
```diff
@@ -341,6 +341,29 @@ public String toString()
...
+     * they terminate.
+     */
+    SOFT_SHUTDOWN,
+    /**
+     * HARD_SHUTDOWN, when this value is set then all operators in the containers
+     * are terminated. If downstream container receives this request before upstream
+     * operator then downstream operators may not process all the data emitted by
+     * upstream operators.
+     */
+    HARD_SHUTDOWN,
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/api/StreamingContainerUmbilicalProtocol.java]: Why is this needed? Killing the application has the same effect.

Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/api/StreamingContainerUmbilicalProtocol.java]: This is for maintaining current behaviour of shutdown, as there is a differece that after SOFT_SHUTDOWN application can not be restarted from last state as the operators are removed from PhysicalPlan. HARD_SHUTDOWN application can be relaunched again.

File: engine/src/main/java/com/datatorrent/stram/api/StreamingContainerUmbilicalProtocol.java [hunk 7]
```diff
@@ -341,6 +341,24 @@ public String toString()
...
+  enum ShutdownType
+  {
+    /**
+     * USER_SHUTDOWN, when this value is set all input operators in the
+     * container terminates and send END_STREAM tuple to downstream
+     * operators. This type of shutdown make sure that data emitted by upstream
+     * operator before shutdown are processed by downstreams operator before
+     * they terminate.
+     */
+    USER_SHUTDOWN,
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/api/StreamingContainerUmbilicalProtocol.java]: I would prefer for these options to reflect the "what" vs. where they are currently used from. How about standard terminology, such as WAIT_TERMINATE for this and IMMEDIATE, ABORT or CANCEL for below?

**Comment by tweise** [comment]
Please check why both CI runs fail.

File: engine/src/main/java/com/datatorrent/stram/api/StreamingContainerUmbilicalProtocol.java (hunk 7 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/api/StreamingContainerUmbilicalProtocol.java]: I have renamed USER_SHUTDOWN to WAIT_TERMINATE and HARD_SHUTDOWN to ABORT.

**Comment by tushargosavi** [comment]
@tweise investigating the CI run fails, I suspect it might be because of APEXCORE-597. investigating it more.

**Comment by tushargosavi** [comment]
@tweise I have resolved CI failures.
